core.network/src/org/netbeans/core/network/proxy/pac/impl/NbPacScriptEvaluator.java

```diff
@@ -0,0 +1,538 @@
+        } catch (ClassNotFoundException ex) {
+            return false;
+        }
+        return true;
+    }
+    
+    private ScriptEngine getNashornJSScriptEngine() {
+        NashornScriptEngineFactory factory = new NashornScriptEngineFactory();
+        return factory.getScriptEngine(new ClassFilterPacHelpers());
+    }
+    
+    private ScriptEngine getGenericJSScriptEngine() {
+        // The result of the statements below may be Rhino, but more likely
+        // - since Java 8 - it will be a Nashorn engine.
+        ScriptEngineManager factory = new ScriptEngineManager();
+        return factory.getEngineByName("JavaScript");
+    }
+    
+    /**
+     * Test if the main entry point, function FindProxyForURL()/FindProxyForURLEx(), 
+     * is available.
+     * 
+     * @param eng
+     * @throws PacParsingException 
+     */
+    private PacJsEntryFunction testScriptEngineNashornJava8u40(ScriptEngine eng) throws PacParsingException {
+        if (isJsFunctionAvailable(eng, PacJsEntryFunction.IPV6_AWARE.getJsFunctionName())) {
+            return PacJsEntryFunction.IPV6_AWARE;
+        }
+        if (isJsFunctionAvailable(eng, PacJsEntryFunction.STANDARD.getJsFunctionName())) {
+            return PacJsEntryFunction.STANDARD;
+        }
+        throw new PacParsingException("Function " + PacJsEntryFunction.STANDARD.getJsFunctionName() + " or " + PacJsEntryFunction.IPV6_AWARE.getJsFunctionName() + " not found in PAC Script.");
+    }
+    
+    private boolean isJsFunctionAvailable(ScriptEngine eng, String functionName) {
+        // We want to test if the function is there, but without actually invoking
+        // it. Don't know how to do this for Rhino or for Nashorn pre Java 8u40.
+        // In other words: The test below only works on Nashorn, Java 8u40 or later.       
+        Object obj = eng.get(functionName);
+        if ( obj != null) {
+            if (obj instanceof ScriptObjectMirror) {
+                    ScriptObjectMirror  som = (ScriptObjectMirror) obj;
+                    if (som.isFunction()) {
+                        return true;
+                    }
+            }
+        }        
+        return false;
+    }
+    
+    private PacJsEntryFunction testScriptEngineGeneric(String pacSource) throws PacParsingException {
+        Pattern pattern1 = Pattern.compile("function\\s+" + PacJsEntryFunction.STANDARD.getJsFunctionName());
+        Pattern pattern2 = Pattern.compile("function\\s+" + PacJsEntryFunction.IPV6_AWARE.getJsFunctionName());
+        if (!pattern2.matcher(pacSource).matches()) {
+            return PacJsEntryFunction.IPV6_AWARE;
+        }
+        if (!pattern1.matcher(pacSource).matches()) {
+            return PacJsEntryFunction.STANDARD;
+        }
+        throw new PacParsingException("Function " + PacJsEntryFunction.STANDARD.getJsFunctionName() + " or " + PacJsEntryFunction.IPV6_AWARE.getJsFunctionName() + " not found in PAC Script.");
```

Comment:
I walked through the code (at least partitally) and here was stopped. I checked rhino and this works ok (`ScriptEngine#get` return non-null):

```java
    public static void main(String[] args) throws Exception {
        ScriptEngineManager m = new ScriptEngineManager();
        ScriptEngine engine = m.getEngineByName("rhino");

        engine.eval("function halloWelt() {\nprint(\"Hallo\");\n}");
        
        System.out.println(engine.get("halloWelt"));
        
        engine.eval("halloWelt()");
    }
```

I think it is a reasonable expectation, that a JS ScriptEngine makes all variables declared in global scope available in the `ScriptContext.ENGINE_SCOPE`. So I basicly suggest to use the `isJsFunctionAvailable` also for the generic case and just skip the `ScriptObjectMirror` check. If someone is stupid enough to declare a variable with the name `FindProxyForURL` or `FindProxyForURLEx` in a PAC script, that is not a function he/she will do other mightily stupid things.

Reply:
Valid point. I'll fix.